sql/distsqlrun: TestDistSQLReadsFillGatewayID failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/97d734c6f9acf9a673ce722ddcec2077f9fc30e6

Parameters:
```
TAGS=
GOFLAGS=
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=834410&tab=buildLog

```
=== RUN   TestDistSQLReadsFillGatewayID
W180814 05:12:45.023772 1494 server/status/runtime.go:287  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180814 05:12:45.062195 1494 server/server.go:824  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180814 05:12:45.064844 1494 server/config.go:492  [n?] 1 storage engine initialized
I180814 05:12:45.064916 1494 server/config.go:495  [n?] RocksDB cache size: 128 MiB
I180814 05:12:45.064933 1494 server/config.go:495  [n?] store 0: in-memory, size 0 B
I180814 05:12:45.081165 1494 server/node.go:373  [n?] **** cluster fafd79fa-52f1-413c-a87b-9123d8b9b1bb has been created
I180814 05:12:45.081202 1494 server/server.go:1399  [n?] **** add additional nodes by specifying --join=127.0.0.1:39509
I180814 05:12:45.081333 1494 gossip/gossip.go:363  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:39509" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:11 > build_tag:"v2.1.0-alpha.20180702-1503-g97d734c" started_at:1534223565081278549 
I180814 05:12:45.082222 1494 storage/store.go:1502  [n1,s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180814 05:12:45.082296 1494 server/node.go:475  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=6.9 KiB), ranges=1, leases=0, queries=0.00, writes=0.00, bytesPerReplica={p10=7043.00 p25=7043.00 p50=7043.00 p75=7043.00 p90=7043.00 pMax=7043.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I180814 05:12:45.082338 1494 storage/stores.go:222  [n1] read 0 node addresses from persistent storage
I180814 05:12:45.082415 1494 server/node.go:696  [n1] connecting to gossip network to verify cluster ID...
I180814 05:12:45.083651 1494 server/node.go:721  [n1] node connected via gossip and verified as part of cluster "fafd79fa-52f1-413c-a87b-9123d8b9b1bb"
I180814 05:12:45.083688 1494 server/node.go:545  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I180814 05:12:45.084065 1494 server/status/recorder.go:652  [n1] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180814 05:12:45.084103 1494 server/server.go:1830  [n1] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180814 05:12:45.084183 1494 server/server.go:1535  [n1] starting https server at 127.0.0.1:46399
I180814 05:12:45.084220 1494 server/server.go:1536  [n1] starting grpc/postgres server at 127.0.0.1:39509
I180814 05:12:45.084232 1494 server/server.go:1537  [n1] advertising CockroachDB node at 127.0.0.1:39509
I180814 05:12:45.093913 1858 server/status/recorder.go:652  [n1,summaries] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180814 05:12:45.110013 1873 sql/event_log.go:126  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I180814 05:12:45.140023 1883 sql/event_log.go:126  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:$1 User:root}
I180814 05:12:45.150179 1727 sql/event_log.go:126  [n1,intExec=disableNetTrace] Event: "set_cluster_setting", target: 0, info: {SettingName:trace.debug.enable Value:false User:root}
I180814 05:12:45.192078 1916 sql/event_log.go:126  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:gen_random_uuid()::STRING User:root}
I180814 05:12:45.197343 1889 sql/event_log.go:126  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I180814 05:12:45.199897 1925 sql/event_log.go:126  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I180814 05:12:45.201448 1494 server/server.go:1619  [n1] done ensuring all necessary migrations have run
I180814 05:12:45.201483 1494 server/server.go:1622  [n1] serving sql connections
I180814 05:12:45.206957 1933 sql/event_log.go:126  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:39509} Attrs: Locality: ServerVersion:2.0-11 BuildTag:v2.1.0-alpha.20180702-1503-g97d734c StartedAt:1534223565081278549} ClusterID:fafd79fa-52f1-413c-a87b-9123d8b9b1bb StartedAt:1534223565081278549 LastUp:1534223565081278549}
I180814 05:12:45.207440 1931 server/server_update.go:67  [n1] no need to upgrade, cluster already at the newest version
W180814 05:12:45.208217 1494 server/status/runtime.go:287  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180814 05:12:45.223657 1494 server/server.go:824  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180814 05:12:45.229609 1494 server/config.go:492  [n?] 1 storage engine initialized
I180814 05:12:45.229634 1494 server/config.go:495  [n?] RocksDB cache size: 128 MiB
I180814 05:12:45.229647 1494 server/config.go:495  [n?] store 0: in-memory, size 0 B
W180814 05:12:45.229805 1494 gossip/gossip.go:1325  [n?] no incoming or outgoing connections
I180814 05:12:45.229897 1494 server/server.go:1401  [n?] no stores bootstrapped and --join flag specified, awaiting init command.
I180814 05:12:45.243171 1988 gossip/client.go:129  [n?] started gossip client to 127.0.0.1:39509
I180814 05:12:45.244528 1901 gossip/server.go:219  [n1] received initial cluster-verification connection from {tcp 127.0.0.1:42929}
I180814 05:12:45.244632 1494 server/node.go:696  [n?] connecting to gossip network to verify cluster ID...
I180814 05:12:45.244663 1494 server/node.go:721  [n?] node connected via gossip and verified as part of cluster "fafd79fa-52f1-413c-a87b-9123d8b9b1bb"
I180814 05:12:45.244937 1998 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180814 05:12:45.245832 1997 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180814 05:12:45.246358 1494 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180814 05:12:45.247048 1494 server/node.go:427  [n?] new node allocated ID 2
I180814 05:12:45.247118 1494 gossip/gossip.go:363  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:42929" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:11 > build_tag:"v2.1.0-alpha.20180702-1503-g97d734c" started_at:1534223565247067163 
I180814 05:12:45.247168 1494 storage/stores.go:222  [n2] read 0 node addresses from persistent storage
I180814 05:12:45.247199 1494 storage/stores.go:241  [n2] wrote 1 node addresses to persistent storage
I180814 05:12:45.248645 1494 server/node.go:671  [n2] bootstrapped store [n2,s2]
I180814 05:12:45.248819 1494 server/node.go:545  [n2] node=2: started with [] engine(s) and attributes []
I180814 05:12:45.249035 1494 server/status/recorder.go:652  [n2] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180814 05:12:45.249064 1494 server/server.go:1830  [n2] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180814 05:12:45.249132 1494 server/server.go:1535  [n2] starting https server at 127.0.0.1:42877
I180814 05:12:45.249152 1494 server/server.go:1536  [n2] starting grpc/postgres server at 127.0.0.1:42929
I180814 05:12:45.249167 1494 server/server.go:1537  [n2] advertising CockroachDB node at 127.0.0.1:42929
I180814 05:12:45.252976 1494 server/server.go:1619  [n2] done ensuring all necessary migrations have run
I180814 05:12:45.253011 1494 server/server.go:1622  [n2] serving sql connections
I180814 05:12:45.256419 2110 server/status/recorder.go:652  [n2,summaries] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
W180814 05:12:45.282961 2006 rpc/nodedialer/nodedialer.go:89  [ct-client] unable to connect to n2: unable to look up descriptor for node 2
I180814 05:12:45.283761 2011 storage/stores.go:241  [n1] wrote 1 node addresses to persistent storage
W180814 05:12:45.291914 1494 server/status/runtime.go:287  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180814 05:12:45.322021 1465 server/server_update.go:67  [n2] no need to upgrade, cluster already at the newest version
I180814 05:12:45.323326 1467 sql/event_log.go:126  [n2] Event: "node_join", target: 2, info: {Descriptor:{NodeID:2 Address:{NetworkField:tcp AddressField:127.0.0.1:42929} Attrs: Locality: ServerVersion:2.0-11 BuildTag:v2.1.0-alpha.20180702-1503-g97d734c StartedAt:1534223565247067163} ClusterID:fafd79fa-52f1-413c-a87b-9123d8b9b1bb StartedAt:1534223565247067163 LastUp:1534223565247067163}
I180814 05:12:45.328924 1494 server/server.go:824  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180814 05:12:45.344341 1494 server/config.go:492  [n?] 1 storage engine initialized
I180814 05:12:45.344369 1494 server/config.go:495  [n?] RocksDB cache size: 128 MiB
I180814 05:12:45.344378 1494 server/config.go:495  [n?] store 0: in-memory, size 0 B
W180814 05:12:45.344457 1494 gossip/gossip.go:1325  [n?] no incoming or outgoing connections
I180814 05:12:45.344522 1494 server/server.go:1401  [n?] no stores bootstrapped and --join flag specified, awaiting init command.
I180814 05:12:45.361123 2219 gossip/client.go:129  [n?] started gossip client to 127.0.0.1:39509
I180814 05:12:45.361446 875 gossip/server.go:219  [n1] received initial cluster-verification connection from {tcp 127.0.0.1:40713}
I180814 05:12:45.362083 1494 server/node.go:696  [n?] connecting to gossip network to verify cluster ID...
I180814 05:12:45.362223 1494 server/node.go:721  [n?] node connected via gossip and verified as part of cluster "fafd79fa-52f1-413c-a87b-9123d8b9b1bb"
I180814 05:12:45.362458 2271 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180814 05:12:45.363278 2270 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180814 05:12:45.363878 1494 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180814 05:12:45.364525 1494 server/node.go:427  [n?] new node allocated ID 3
I180814 05:12:45.364590 1494 gossip/gossip.go:363  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:40713" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:11 > build_tag:"v2.1.0-alpha.20180702-1503-g97d734c" started_at:1534223565364534404 
I180814 05:12:45.364633 1494 storage/stores.go:222  [n3] read 0 node addresses from persistent storage
I180814 05:12:45.364665 1494 storage/stores.go:241  [n3] wrote 2 node addresses to persistent storage
I180814 05:12:45.365484 2280 storage/stores.go:241  [n1] wrote 2 node addresses to persistent storage
I180814 05:12:45.365869 2281 storage/stores.go:241  [n2] wrote 2 node addresses to persistent storage
I180814 05:12:45.366452 1494 server/node.go:671  [n3] bootstrapped store [n3,s3]
I180814 05:12:45.366572 1494 server/node.go:545  [n3] node=3: started with [] engine(s) and attributes []
I180814 05:12:45.366868 1494 server/status/recorder.go:652  [n3] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180814 05:12:45.366897 1494 server/server.go:1830  [n3] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180814 05:12:45.366968 1494 server/server.go:1535  [n3] starting https server at 127.0.0.1:43147
I180814 05:12:45.366982 1494 server/server.go:1536  [n3] starting grpc/postgres server at 127.0.0.1:40713
I180814 05:12:45.367006 1494 server/server.go:1537  [n3] advertising CockroachDB node at 127.0.0.1:40713
I180814 05:12:45.368072 2362 server/status/recorder.go:652  [n3,summaries] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180814 05:12:45.393406 1494 server/server.go:1619  [n3] done ensuring all necessary migrations have run
I180814 05:12:45.393439 1494 server/server.go:1622  [n3] serving sql connections
I180814 05:12:45.409679 2398 sql/event_log.go:126  [n3] Event: "node_join", target: 3, info: {Descriptor:{NodeID:3 Address:{NetworkField:tcp AddressField:127.0.0.1:40713} Attrs: Locality: ServerVersion:2.0-11 BuildTag:v2.1.0-alpha.20180702-1503-g97d734c StartedAt:1534223565364534404} ClusterID:fafd79fa-52f1-413c-a87b-9123d8b9b1bb StartedAt:1534223565364534404 LastUp:1534223565364534404}
I180814 05:12:45.411937 2396 server/server_update.go:67  [n3] no need to upgrade, cluster already at the newest version
I180814 05:12:45.427142 2422 sql/event_log.go:126  [n1,client=127.0.0.1:55480,user=root] Event: "create_database", target: 52, info: {DatabaseName:test Statement:CREATE DATABASE IF NOT EXISTS test User:root}
I180814 05:12:45.431006 2422 sql/event_log.go:126  [n1,client=127.0.0.1:55480,user=root] Event: "create_table", target: 53, info: {TableName:test.public.t Statement:CREATE TABLE test.public.t (num INT PRIMARY KEY) User:root}
I180814 05:12:45.433998 2422 storage/replica_command.go:280  [n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key /Table/53/1/1 [r2]
I180814 05:12:45.446448 2422 storage/replica_command.go:280  [n1,s1,r2/1:/{Table/53/1/1-Max}] initiating a split of this range at key /Table/53/1/2 [r3]
I180814 05:12:45.451143 2422 storage/replica_command.go:280  [n1,s1,r3/1:/{Table/53/1/2-Max}] initiating a split of this range at key /Table/53/1/3 [r4]
I180814 05:12:45.471339 2422 storage/replica_raftstorage.go:521  [n1,s1,r2/1:/Table/53/1/{1-2}] generated preemptive snapshot 3ae2b104 at index 17
--- FAIL: TestDistSQLReadsFillGatewayID (0.46s)
	cluster_test.go:569: pq: while adding target n2,s2: snapshot failed: unable to dial n2: breaker open

```

--- a/pkg/server/testserver.go
+++ b/pkg/server/testserver.go
@@ -25,6 +25,7 @@ import (
 
 	"github.com/pkg/errors"
 
+	"github.com/cenk/backoff"
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/config"
 	"github.com/cockroachdb/cockroach/pkg/gossip"
@@ -48,6 +49,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/metric"
 	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/rubyist/circuitbreaker"
 )
 
 const (
@@ -336,6 +338,14 @@ func (ts *TestServer) Start(params base.TestServerArgs) error {
 	ts.Server, err = NewServer(*ts.Cfg, params.Stopper)
 	if err != nil {
 		return err
+	}
+
+	// Create a breaker which never trips and never backs off to avoid
+	// introducing timing-based flakes.
+	ts.rpcContext.BreakerFactory = func() *circuit.Breaker {
+		return circuit.NewBreakerWithOptions(&circuit.Options{
+			BackOff: &backoff.ZeroBackOff{},
+		})
 	}
 
 	// Our context must be shared with our server.

